File: packages/platform-ios/Sources/Keyward/Keyward.swift
public final class Keyward {
    static var backend: KeychainBackend = SystemKeychainBackend()
    private static let registry = KeyRegistry()

    // MARK: - User ID management (uses internal registry)

    public static func setUserId(_ userId: String) {
        registry.setUserId(userId)
    }

    public static func clearUserId() {
        registry.clearUserId()
    }

    public static func getUserId() -> String? {
        registry.getUserId()
    }

    // MARK: - High-level API with internal registry

    public static func get(_ keyDef: KeyDef) throws -> String? {
        try get(keyDef, registry: registry)
    }

    public static func set(_ keyDef: KeyDef, value: String) throws {
        try set(keyDef, value: value, registry: registry)
    }

    public static func remove(_ keyDef: KeyDef) throws {
        try remove(keyDef, registry: registry)
    }

    // MARK: - High-level API with explicit registry

    public static func get(_ keyDef: KeyDef, registry: KeyRegistry) throws -> String? {
        let resolvedKey = try registry.resolve(keyDef)
        return getRaw(key: resolvedKey)
    }

    public static func set(_ keyDef: KeyDef, value: String, registry: KeyRegistry) throws {
        let resolvedKey = try registry.resolve(keyDef)
        try setRaw(key: resolvedKey, value: value)
    }

    public static func remove(_ keyDef: KeyDef, registry: KeyRegistry) throws {
        let resolvedKey = try registry.resolve(keyDef)
        removeRaw(key: resolvedKey)
    }

    public static func keys() -> [String] {
        backend.allKeys()
    }

    public static func wipeUser(_ userId: String) {
        let prefix = KeyRegistry.getUserPrefix(userId)
        for key in keys() where key.hasPrefix(prefix) {
            removeRaw(key: key)
        }
    }

    // MARK: - Raw API (Capacitor bridge)

    public static func getRaw(key: String) -> String? {
        backend.get(key: key)
    }

    public static func setRaw(key: String, value: String) throws {
        try backend.set(key: key, value: value)
    }

    public static func removeRaw(key: String) {
        backend.remove(key: key)
    }

    public static func clearAll() {
        backend.clear()
    }

    private init() {}
}

File: packages/platform-android/src/main/java/org/keyward/Keyward.java
package org.keyward;

import java.util.ArrayList;
import java.util.List;

public final class Keyward {
    static KeystoreBackend backend;
    private static final KeyRegistry registry = new KeyRegistry();

    public static void setBackend(KeystoreBackend b) {
        backend = b;
    }

    // User ID management (uses internal registry)

    public static void setUserId(String userId) {
        registry.setUserId(userId);
    }

    public static void clearUserId() {
        registry.clearUserId();
    }

    public static String getUserId() {
        return registry.getUserId();
    }

    // High-level API with internal registry

    public static String get(KeyDef keyDef) throws KeywardException {
        return get(keyDef, registry);
    }

    public static void set(KeyDef keyDef, String value) throws KeywardException {
        set(keyDef, value, registry);
    }

    public static void remove(KeyDef keyDef) throws KeywardException {
        remove(keyDef, registry);
    }

    // High-level API with explicit registry

    public static String get(KeyDef keyDef, KeyRegistry registry) throws KeywardException {
        String resolvedKey = registry.resolve(keyDef);
        return getRaw(resolvedKey);
    }

    public static void set(KeyDef keyDef, String value, KeyRegistry registry) throws KeywardException {
        String resolvedKey = registry.resolve(keyDef);
        setRaw(resolvedKey, value);
    }

    public static void remove(KeyDef keyDef, KeyRegistry registry) throws KeywardException {
        String resolvedKey = registry.resolve(keyDef);
        removeRaw(resolvedKey);
    }

    public static List<String> keys() {
        return backend.allKeys();
    }

    public static void wipeUser(String userId) {
        String prefix = KeyRegistry.getUserPrefix(userId);
        for (String key : new ArrayList<>(keys())) {
            if (key.startsWith(prefix)) {
                removeRaw(key);
            }
        }
    }

    // Raw API (Capacitor bridge)

    public static String getRaw(String key) {
        return backend.get(key);
    }

    public static void setRaw(String key, String value) throws KeywardException {
        backend.set(key, value);
    }

    public static void removeRaw(String key) {
        backend.remove(key);
    }

    public static void clearAll() {
        backend.clear();
    }

    private Keyward() {}
}
